src/template-expr.adoc

```diff
@@ -0,0 +1,255 @@
+[[sec:templexpr]]
+== Template Expressions
+
+include::styles.adoc[]
+
+The behavior of a macro is defined in terms of an expression language.
+Like encoding directives and modules, this language is expressed as Ion data, and
+the meaning of templates is defined structurally and recursively based on the Ion data model.
+
+
+=== Grammar
+
+Here's the relevant portion of the <<sec:grammar,domain grammar>>:
+
+include::grammar.adoc[tag=templexpr]
+
+An expression in this language is called a _template_, and the expansion of a template (that is,
+its evaluation) produces a stream of Ion values.
+The central design concept is that symbols denote variable references, S-expressions denote
+operator invocations, and other Ion types denote values of that type.
+
+
+==== Symbols are Variable References
+
+When a template is an Ion symbol, it denotes a reference to a variable, either a macro parameter
+or a local binding from for.
+The result of this template is the stream of values produced by that variable.
+
+The symbols used for variable names must be _identifiers_ as defined by the Ion specification:
+a sequence of ASCII letters, digits, or the characters `$` (dollar sign) or `_`
+(underscore), not starting with a digit.
+
+When a template is expected, the symbols `$0` and `null.symbol` evoke a syntax error, as does any
+annotated symbol.
+
+TIP: To denote the literal symbol `foo`, use the template `(*literal* foo)`.
+
+
+==== Other Scalars are Literals
+
+When a template is a non-symbol Ion scalar, it denotes a literal value, and the template expands
+into that value. Any annotations on the template are included in the output.
+
+
+==== Lists and Structs are Quasi-Literals
+
+When a template is an Ion list or struct, it denotes a quasi-literal of the same type.
+We say “quasi” literal because the elements of the container are treated as templates, not
+literal values.
+
+When a template is a list, it expands into a list with the same annotations.
+The elements of the list-template are each treated as templates themselves.
+Each sub-template may produce any number of values, and the resulting streams are all
+concatenated to produce the output list.
+
+----
+[1, [2, 3], 4]       ⇒ [1, [2, 3], 4]
+[1, (values 2 3), 4] ⇒ [1, 2, 3, 4]
+[1, (values), 3]     ⇒ [1, 3]
+----
+
+When a template is a struct, it expands into a struct with the same annotations.
+The struct-template’s field names are treated as literals, and field values are treated as
+sub-templates, and the output struct contains the given names and their associated sub-template
+expansions.
+
+Field-value sub-templates MAY produce multiple values.
+When a sub-template produces more than one result, then the output struct will have more than one
+field with the same name.
+When a sub-template produces no results, then nothing is added to the output.
+
+----
+{a:(values 1 2)} ⇒ {a:1, a:2}   // or, equivalently, {a:2, a:1}
+{f:(values)}     ⇒ {}
+----
+
```

Comment:
I'm not sure what you mean.  I don't think that the template language has a way to splice a struct into a containing struct, because there's no way to omit the field-name from the syntax.

If you are referring to the "fanout" on L73, that's explained on L68.